Add autograd support for square (including complex) + Sparse CSR support

krshrimali opened the pull request:
This PR is a follow-up PR from https://github.com/pytorch/pytorch/pull/69292, to add derivative support for `square`. This also adds support automatic differentiation for outputs with complex dtypes, along with `square` now supporting Sparse CSR layout.

- ~I've skipped forward mode AD tests, for now, will have to take a look at how to add that for `square`. Hopefully in another PR.~ Forward mode AD support has been added as well.
- ~Instead of writing `2 * grad * self.conj()`, I thought I'll just use `pow_backward(grad, self, 2)` in the formula. I assume both are the same? (please correct if I'm wrong)~ Used the formula instead, based on suggestion: https://github.com/pytorch/pytorch/pull/69510#issuecomment-987693858.

Note: Once #69292 is merged, this PR will be limited to the changes made for `square`. For CI to run, I've added changes from #69292 in this PR as well. To make the review easy, I've marked the relevant changes by review comments. Thanks!

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

krshrimali committed:
Sparse CSR Tensors, use macros, and add support for other unary ufuncs

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into sparse/csr/unary_ops

krshrimali committed:
abs: add entry in native and enable tests for sparse CSR

krshrimali committed:
Handle aliases

krshrimali committed:
Handle aliases

krshrimali committed:
Add more ops, remove absolute (abs is there)

krshrimali committed:
conj_physical support for sparse CSR, conj - doesn't have an out variant :/

krshrimali committed:
Rest of the ops, refined, and... almost done

krshrimali committed:
Final ops

krshrimali committed:
not working with positive

krshrimali committed:
Add dispatches for some ops

krshrimali committed:
Merge conflicts resolved

krshrimali committed:
Tests for out and inplace variants

krshrimali committed:
abs_ does not support complex inputs

krshrimali committed:
Revise implementation for inplace versions

krshrimali committed:
Remove separate inplace impl of abs

krshrimali committed:
Minor revision in comment

krshrimali committed:
bring back dispatch for square

krshrimali committed:
complex support for square (derivative), sparse csr support added

krshrimali commented on tools/autograd/gen_variable_type.py:
```diff
@@ -89,7 +89,7 @@
     '_conj', 'sin', 'cos', 'mul', 'sinc', 'sinh', 'cosh', '__rmul__',
     'sgn', 'asin', 'acos', 'sub', 'div', 'cat', 'view_as_complex',
-    'neg', 'complex', 'select', '_s_where', 'as_strided', 'slice', 'constant_pad_nd',
+    'neg', 'complex', 'select', '_s_where', 'as_strided', 'slice', 'square', 'constant_pad_nd',
```
relevant change

krshrimali commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -11624,6 +11645,9 @@ def ref_pairwise_distance(input1, input2):
                        # Reference: https://github.com/pytorch/pytorch/pull/52551#issuecomment-782596181
                        DecorateInfo(unittest.skip("Skipped!"), 'TestUnaryUfuncs', 'test_reference_numerics_hard',
                                     dtypes=[torch.bfloat16]),
+                       # Forward Mode AD support not added yet for square
+                       DecorateInfo(unittest.skip("Skipped!"), 'TestGradients', 'test_inplace_forward_mode_AD'),
+                       DecorateInfo(unittest.skip("Skipped!"), 'TestGradients', 'test_forward_mode_AD'),
```
relevant change

IvanYashchuk replied:
What is the error here? You've included `result: auto_element_wise` in the `derivatives.yaml` so forward AD should be working.

krshrimali commented on tools/autograd/derivatives.yaml:
```diff
@@ -1222,6 +1222,10 @@
   self: -0.5 * grad * result.pow(3).conj()
   result: auto_element_wise
 
+- name: square(Tensor self) -> Tensor
+  self: pow_backward(grad, self, 2)
+  result: auto_element_wise
+
```
relevant change

krshrimali commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -11608,6 +11628,7 @@ def ref_pairwise_distance(input1, input2):
                    dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
                    decorators=(precisionOverride({torch.complex64: 3e-4, torch.bfloat16: 3e-1}),),
                    supports_forward_ad=True,
+                   supports_sparse_csr=True,
```
relevant change

krshrimali commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -164,17 +173,70 @@ bool is_square_or_vec(int64_t dim_i, int64_t dim_j, int64_t dim_k) {
+CREATE_UNARY_UFUNC(sinh);
+CREATE_UNARY_UFUNC(sgn);
+CREATE_UNARY_UFUNC(sqrt);
+CREATE_UNARY_UFUNC(square);
```
relevant change

krshrimali commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4283,18 +4336,26 @@
 - func: square(Tensor self) -> Tensor
   device_check: NoCheck   # TensorIterator
   variants: function, method
+  dispatch:
+    CPU, CUDA: square
+    SparseCsrCPU, SparseCsrCUDA: square_sparse_csr
 
 - func: square_(Tensor(a!) self) -> Tensor(a!)
   device_check: NoCheck   # TensorIterator
   variants: function, method
+  dispatch:
+    CPU, CUDA: square_
+    SparseCsrCPU, SparseCsrCUDA: square_sparse_csr_
 
 - func: square.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
   dispatch:
     CPU, CUDA: square_out
+    SparseCsrCPU, SparseCsrCUDA: square_sparse_csr_out
```
relevant change

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into sparse/csr/square

Lezcano commented:
Please don't assume the backward for pow is the same as the formula. If the formula is simple, please write it inline, as that way we know that it's as fast as it can be.
You added forward AD support via `the `auto_element_wise` keyword. That should be enough. What's the errors that you get from the tests?

krshrimali committed:
Use formula instead of pow_backward, forward mode AD was added, remove skips

IvanYashchuk reviewed (commented):
`torch.square` (`at::square`) is a composite function. There's no need on specializing the dispatcher for this. Unfortunately, it's also a breaking change to use 
```
  dispatch:
    CPU, CUDA: square
```
To get `torch.square` working for sparse csr tensor correctly `torch.pow` needs to be implemented.

krshrimali commented:
> Please don't assume the backward for pow is the same as the formula. If the formula is simple, please write it inline, as that way we know that it's as fast as it can be.

Sure, makes sense. Should be fixed now.

> You added forward AD support via `the `auto_element_wise` keyword. That should be enough. What's the errors that you get from the tests?

Ah, the failures were before I added `auto_element_wise`. I've removed the skips now. Thanks!

krshrimali committed:
Remove square dispatches

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into sparse/csr/square

IvanYashchuk commented:
This PR is closed in favor of https://github.com/pytorch/pytorch/pull/69603.
